connmgr: switch to using net.Addr interface throughout for addresses

Roasbeef opened the pull request:
This commit modifies the ConnManager to use the `net.Addr interface through the package instead of a plain string to represent and manipulate addresses. This change makes the package much more general as users of the package can possibly utilize custom implementations of the net.Addr interface to establish connections.

More precisely, the `ConnReq` struct has been modified to use a net.Addr instance explicitly, and the `DialFunc` type has also been modified to take a `net.Addr` directly. This latter change gives functions that adhere to the `DialFunc` type more flexibility as to exactly how the connection is established.

My motivation for this change is that I'd like to use the `connmgr` package within [`lnd`](https://github.com/lightningnetwork/lnd) to manage attempting to establish persistent outbound connections to channel counter parties. Within the Lightning Network, [all p2p connections are encrypted+authenticated by default](https://github.com/lightningnetwork/lnd/blob/master/brontide/noise.go). 

Before this commit, this additional constraint meant that I couldn't directly use the `connmgr` because all the functions/structs referred to addresses explicitly by a string. This is rather limiting as in LN addresses also are tied to a long-term identity public key. By modifying the package to use a `net.Addr` throughout, I'm utilize our implementation of `net.Addr` whose concrete struct contains the additional information required to establish an encrypted connection to peers on the network.

davecgh commented:
I haven't looked at this super closely yet, but a couple of initial comments:

1) I generally like the 
2) I'm fairly certain the current implementation will leak the DNS resolution over Tor because it's calling `net.ResolveTCPAddr` instead of using `cfg.lookup` and `cfg.onionlookup` which are setup in `config.go` based upon the tor parameters.

Roasbeef commented:
Excellent observation. I've pushed a rebased and slightly modified version of the original commit. Rather than using `net.ResolveTCPAddr` everywhere due to the change, the `connmgr.Config` struct has gained a new attribute: a closure for performing DNS lookups. This is used throughout rather than potentially incorrectly performing DNS lookup as you detailed.

davecgh commented on connmgr/connmanager.go:
```diff
@@ -288,7 +292,30 @@ func (cm *ConnManager) NewConnReq() {
 	}
-	c.Addr = addr
+
+	host, strPort, err := net.SplitHostPort(addr)
```
Since everything else in the package now deals with `net.Addr`, I think this should be modified instead such that the `GetNewAddress` callback returns a `net.Addr` and then the `newAddressFunc` closure in `server.go` which is provided would pick up the parsing, lookup, and `net.Addr` creation.  This would work as is, but it seems to go against the spirit of the PR which seems to be aimed at making it agnostic to the specific address type.  In fact, I believe that would get rid of the need to even pass in a lookup function at all and allow the code you added in `server.go` under the `connectNodeMsg` case and `newServer` function to be refactored into a common code used in all three places (`connectNodeMsg`, `newServer`, and `newAddressFunc`).  What do you think?

I expect `addrmgr.GetAddress` itself should probably also be modified at some point to return `net.Addr` too, but that can be saved for another PR.

Roasbeef replied:
Great suggestion! Modifying `GetNewAddress` to return a `net.Addr` directly allows for a good bit of code consolidation and allows this PR to keep the bulk of the parsing/lookup logic in a single area. 

I've just pushed a commit which makes the modification you mentioned in your comment above. 

> I expect addrmgr.GetAddress itself should probably also be modified at some point to return net.Addr too, but that can be saved for another PR.

One way to implement this would be to make `wire.NetAddress` adhere to the `net.Addr` interface by adding a `.Network()` method. This would allow the addresses returned by `addrmgr.GetAddress` to be directly passed into this context. 

As an aside, I think the `addrmgr` packaged is currently one of the most neglected packages within the project. The code could use a bit of cleaning up, and we could also dump the json file dump format in favor of creating a dedicated bucket for the data within the database. Additionally, it'd be nice to implement some of the counter measures for the attacks described [here](https://www.usenix.org/system/files/conference/usenixsecurity15/sec15-paper-heilman.pdf) and [here](https://arxiv.org/abs/1605.07524).

davecgh replied:
Nice.  Yes, I completely agree about `addrmgr`.  I've mentioned a few times as well that it's probably the package that needs the most improvement/cleanup in the entire repository.

That seems like a reasonable idea to make it implement `net.Addr`.  Another approach would be to create a new type in addrmgr which embeds a `wire.NetAddress` and adds the `Network` method so it implements `net.Addr`.  That way the wire type wouldn't have to be directly modified and additional metadata could be added to the new type if desired as well.

Also, there is an open issue (#581) for converting it to make use of the database as you mentioned.

davecgh commented on server.go:
```diff
@@ -2432,9 +2438,13 @@ func newServer(listenAddrs []string, db database.DB, chainParams *chaincfg.Param
-				return addrmgr.NetAddressKey(addr.NetAddress()), nil
+
+				addrString := addrmgr.NetAddressKey(addr.NetAddress())
+
```
Super minor nit: This blank line probably shouldn't be here since it's grouped with the return in terms of logic.

Roasbeef replied:
Fixed.

Roasbeef commented:
Thanks! I've addressed your final review item and squashed down the review commits into the original PR commit. I've also updated the commit message to reflect the new approach taken w.r.t `GetNewAddress`.
